File: src/filters/prescriptive.ts
export interface FilterResult {
  passed: boolean;
  violations: string[];
  rewritten?: string;
}

// Patterns that indicate prescriptive/directive language
const PRESCRIPTIVE_PATTERNS: [RegExp, string][] = [
  [/\byou\s+should\b/gi, 'you should'],
  [/\byou\s+must\b/gi, 'you must'],
  [/\byou\s+need\s+to\b/gi, 'you need to'],
  [/\byou\s+have\s+to\b/gi, 'you have to'],
  [/\bdo\s+this\b/gi, 'do this'],
  [/\bdon't\s+do\b/gi, "don't do"],
  [/\bthe\s+best\s+(option|choice|way)\s+is\b/gi, 'the best X is'],
  [/\bhere's\s+what\s+you\s+(must|should|need\s+to)\s+do\b/gi, "here's what you must do"],
  [/\bthe\s+right\s+(answer|choice|decision)\s+is\b/gi, 'the right X is'],
  [/\balways\s+do\b/gi, 'always do'],
  [/\bnever\s+do\b/gi, 'never do'],
  [/\bi\s+recommend\s+that\s+you\b/gi, 'I recommend that you'],
  [/\bmy\s+advice\s+is\s+to\b/gi, 'my advice is to'],
];

// Replacement suggestions for rewrites
const REWRITES: Record<string, string> = {
  'you should': 'you might consider',
  'you must': 'one approach could be to',
  'you need to': 'it may help to',
  'you have to': 'one option is to',
  'the best X is': 'one possibility is',
  'the right X is': 'depending on your situation,',
  'always do': 'in many cases,',
  'never do': 'it may be worth reconsidering',
  'I recommend that you': 'some people find it helpful to',
  'my advice is to': 'one perspective is that',
};

export function filterPrescriptive(output: string): FilterResult {
  const violations: string[] = [];
  let rewritten = output;

  for (const [pattern, label] of PRESCRIPTIVE_PATTERNS) {
    if (pattern.test(output)) {
      violations.push(label);
      // Apply rewrite if available
      const replacement = REWRITES[label];
      if (replacement) {
        rewritten = rewritten.replace(pattern, replacement);
      }
    }
  }

  return {
    passed: violations.length === 0,
    violations,
    rewritten: violations.length > 0 ? rewritten : undefined
  };
}

File: src/postfilters/prescriptive.py
"""
Prescriptive Language Filter

Blocks/rewrites authoritative "you should" patterns.
"""

import re
from typing import List


# Use late import to avoid circular dependency
def _get_result_class():
    from . import PostfilterResult, PostfilterOutcome
    return PostfilterResult, PostfilterOutcome


PRESCRIPTIVE_PATTERNS = [
    (r"(?i)\byou\s+should\b", "you should"),
    (r"(?i)\byou\s+must\b", "you must"),
    (r"(?i)\bthe\s+best\s+option\s+is\b", "the best option is"),
    (r"(?i)\bdo\s+this\b", "do this"),
    (r"(?i)\bI\s+recommend\b", "I recommend"),
    (r"(?i)\byou\s+need\s+to\b", "you need to"),
    (r"(?i)\bI\s+advise\b", "I advise"),
]

REPLACEMENTS = {
    "you should": "you might consider",
    "you must": "it may help to",
    "the best option is": "one effective approach is",
    "do this": "consider this",
    "I recommend": "one option is",
    "you need to": "you could",
    "I advise": "a possible path is",
}


def check_prescriptive(output: str, mode: str = "TRANSACTIONAL"):
    """
    Check for prescriptive language and rewrite if found.
    
    In PLAY mode, prescriptive language is allowed.
    """
    from . import PostfilterResult, PostfilterOutcome
    
    if mode == "PLAY":
        return PostfilterResult(
            outcome=PostfilterOutcome.ALLOWED,
            output=output,
            filter_name="prescriptive"
        )
    
    violations = []
    current = output
    
    for pattern, name in PRESCRIPTIVE_PATTERNS:
        if re.search(pattern, current):
            violations.append(f"prescriptive:{name}")
    
    if violations:
        # Apply replacements
        rewritten = output
        for pattern, name in PRESCRIPTIVE_PATTERNS:
            if name in REPLACEMENTS:
                # Case-insensitive replacement
                replacement = REPLACEMENTS[name]
                rewritten = re.sub(
                    pattern, 
                    lambda m: replacement if m.group()[0].islower() 
                        else replacement[0].upper() + replacement[1:],
                    rewritten,
                    count=0
                )
        
        return PostfilterResult(
            outcome=PostfilterOutcome.REWRITTEN,
            output=rewritten,
            filter_name="prescriptive",
            violations=violations
        )
    
    return PostfilterResult(
        outcome=PostfilterOutcome.ALLOWED,
        output=output,
        filter_name="prescriptive"
    )
